File: backend/app/journal.py
"""Journal service — server-side Supabase client. Mirrors src/journal.js so
the memory format and schema stay compatible with existing frontend writes.

The frontend can continue writing directly with the anon key while the
backend also reads/writes with the service key — both paths use the same
`analyses` table and produce identical memory blocks."""
import logging
import re
from typing import Any, Dict, List, Optional

from supabase import Client, create_client

from .config import get_settings

log = logging.getLogger("desk.journal")

# Keep in sync with MISTAKE_CATEGORIES in src/journal.js.
MISTAKE_CATEGORIES = [
    "stale_data",
    "wrong_assumption",
    "valuation_error",
    "earnings_surprise",
    "macro_shift",
    "thesis_drift",
    "data_gap",
    "concentration",
    "other",
]

PROMPT_VERSION = "v2-2026-07"

_client: Optional[Client] = None
_client_key = ""


def _get_client() -> Optional[Client]:
    """Lazy singleton — rebuilds only if config changes."""
    global _client, _client_key
    s = get_settings()
    if not s.supabase_url or not s.supabase_service_key:
        return None
    key = f"{s.supabase_url}|{s.supabase_service_key}"
    if _client is not None and _client_key == key:
        return _client
    _client = create_client(s.supabase_url, s.supabase_service_key)
    _client_key = key
    return _client


def is_ready() -> bool:
    return _get_client() is not None


def _parse_price(v: Any) -> Optional[float]:
    if v is None:
        return None
    if isinstance(v, (int, float)):
        return float(v) if v == v else None  # reject NaN
    s = re.sub(r"[,\s₹$€£]", "", str(v))
    s = re.sub(r"^Rs\.?", "", s, flags=re.IGNORECASE)
    try:
        f = float(s)
        return f if f == f else None
    except (ValueError, TypeError):
        return None


def save_analysis(record: Dict[str, Any]) -> Dict[str, Any]:
    c = _get_client()
    if c is None:
        return {"ok": False, "reason": "not-configured"}
    row = {
        "ticker": record.get("ticker"),
        "company_name": record.get("company_name") or record.get("companyName"),
        "provider": record.get("provider"),
        "horizon": record.get("horizon"),
        "price_at_analysis": _parse_price(record.get("price_at_analysis") or record.get("priceAtAnalysis")),
        "currency": record.get("currency") or "INR",
        "recommendation": record.get("recommendation"),
        "confidence": _to_int(record.get("confidence")),
        "risk": record.get("risk"),
        "fair_value": record.get("fair_value") or record.get("fairValue"),
        "buy_range": record.get("buy_range") or record.get("buyRange"),
        "target_price": _parse_price(record.get("fair_value") or record.get("fairValue")),
        "summary": record.get("summary"),
        "reasoning": record.get("reasoning"),
        "bull_case": record.get("bull_case") or record.get("bullCase"),
        "bear_case": record.get("bear_case") or record.get("bearCase"),
        "full_data": record.get("full_data") or record.get("fullData"),
        # v2 fields
        "record_type": record.get("record_type") or record.get("recordType") or "stock_deepdive",
        "market_regime": record.get("market_regime") or record.get("marketRegime"),
        "model_version": record.get("model_version") or record.get("modelVersion"),
        "prompt_version": record.get("prompt_version") or record.get("promptVersion") or PROMPT_VERSION,
        "grounded": record.get("grounded"),
        "stated_confidence": _to_int(record.get("stated_confidence") or record.get("statedConfidence")),
        "calibrated_confidence": _to_int(record.get("calibrated_confidence") or record.get("calibratedConfidence")),
        "invalidators": record.get("invalidators"),
        "recheck_triggers": record.get("recheck_triggers") or record.get("recheckTriggers"),
        "prior_mistakes_considered": record.get("prior_mistakes_considered") or record.get("priorMistakesConsidered"),
        "safety_gate": record.get("safety_gate") or record.get("safetyGate"),
    }
    try:
        resp = c.table("analyses").insert(row).execute()
        rid = (resp.data or [{}])[0].get("id")
        log.info("journal.save.ok", extra={"ticker": row["ticker"], "record_type": row["record_type"], "id": rid})
        return {"ok": True, "id": rid}
    except Exception as e:
        log.warning("journal.save.fail", extra={"reason": str(e)})
        return {"ok": False, "reason": str(e)}


def _to_int(v: Any) -> Optional[int]:
    if v is None:
        return None
    try:
        return int(float(v))
    except (ValueError, TypeError):
        return None


def list_for_ticker(ticker: str, limit: int = 5) -> List[Dict[str, Any]]:
    c = _get_client()
    if c is None or not ticker:
        return []
    try:
        r = (
            c.table("analyses")
            .select("*")
            .eq("ticker", ticker.upper())
            .order("created_at", desc=True)
            .limit(limit)
            .execute()
        )
        return r.data or []
    except Exception as e:
        log.warning("journal.list_for_ticker.fail", extra={"reason": str(e)})
        return []


def list_recent_misses(limit: int = 5) -> List[Dict[str, Any]]:
    c = _get_client()
    if c is None:
        return []
    try:
        r = (
            c.table("analyses")
            .select("*")
            .eq("outcome", "MISS")
            .order("return_pct", desc=False)  # most negative first
            .limit(limit)
            .execute()
        )
        return r.data or []
    except Exception as e:
        log.warning("journal.list_recent_misses.fail", extra={"reason": str(e)})
        return []


def list_misses_for_ticker(ticker: str, limit: int = 3) -> List[Dict[str, Any]]:
    c = _get_client()
    if c is None or not ticker:
        return []
    try:
        r = (
            c.table("analyses")
            .select("*")
            .eq("ticker", ticker.upper())
            .eq("outcome", "MISS")
            .order("created_at", desc=True)
            .limit(limit)
            .execute()
        )
        return r.data or []
    except Exception:
        return []


def list_all(limit: int = 100, record_type: Optional[str] = None) -> List[Dict[str, Any]]:
    c = _get_client()
    if c is None:
        return []
    try:
        q = c.table("analyses").select("*").order("created_at", desc=True).limit(limit)
        if record_type:
            q = q.eq("record_type", record_type)
        return (q.execute()).data or []
    except Exception as e:
        log.warning("journal.list_all.fail", extra={"reason": str(e)})
        return []


def update_review(analysis_id: str, patch: Dict[str, Any]) -> Dict[str, Any]:
    c = _get_client()
    if c is None:
        return {"ok": False, "reason": "not-configured"}
    from datetime import datetime, timezone
    row: Dict[str, Any] = {"reviewed_at": datetime.now(timezone.utc).isoformat()}
    if "priceAtReview" in patch or "price_at_review" in patch:
        row["price_at_review"] = _parse_price(patch.get("price_at_review") or patch.get("priceAtReview"))
    if patch.get("outcome"):
        row["outcome"] = patch["outcome"]
    if "returnPct" in patch or "return_pct" in patch:
        v = patch.get("return_pct") or patch.get("returnPct")
        row["return_pct"] = float(v) if v is not None else None
    if "lessons" in patch:
        row["lessons"] = patch["lessons"] or None
    if "mistakeCategory" in patch or "mistake_category" in patch:
        row["mistake_category"] = patch.get("mistake_category") or patch.get("mistakeCategory") or None
    if "whatWasMissed" in patch or "what_was_missed" in patch:
        row["what_was_missed"] = patch.get("what_was_missed") or patch.get("whatWasMissed") or None
    if "whatToCheck" in patch or "what_to_check" in patch:
        row["what_to_check"] = patch.get("what_to_check") or patch.get("whatToCheck") or None
    if "marketRegime" in patch or "market_regime" in patch:
        row["market_regime"] = patch.get("market_regime") or patch.get("marketRegime") or None
    try:
        c.table("analyses").update(row).eq("id", analysis_id).execute()
        return {"ok": True}
    except Exception as e:
        return {"ok": False, "reason": str(e)}


def delete_analysis(analysis_id: str) -> Dict[str, Any]:
    c = _get_client()
    if c is None:
        return {"ok": False, "reason": "not-configured"}
    try:
        c.table("analyses").delete().eq("id", analysis_id).execute()
        return {"ok": True}
    except Exception as e:
        return {"ok": False, "reason": str(e)}


def get_mistake_stats(limit: int = 500) -> Optional[Dict[str, Any]]:
    """Client-side aggregation over recent rows — fine for single-user scale."""
    c = _get_client()
    if c is None:
        return None
    try:
        r = (
            c.table("analyses")
            .select("recommendation,outcome,return_pct,mistake_category,record_type")
            .order("created_at", desc=True)
            .limit(limit)
            .execute()
        )
        rows = r.data or []
    except Exception:
        return None

    by_action: Dict[str, Dict[str, Any]] = {}
    by_mistake: Dict[str, int] = {}
    for row in rows:
        a = row.get("recommendation") or "UNKNOWN"
        st = by_action.setdefault(a, {"hit": 0, "miss": 0, "pending": 0, "returns": []})
        outcome = row.get("outcome")
        if outcome == "HIT":
            st["hit"] += 1
        elif outcome == "MISS":
            st["miss"] += 1
        else:
            st["pending"] += 1
        rp = row.get("return_pct")
        if rp is not None:
            try:
                st["returns"].append(float(rp))
            except (ValueError, TypeError):
                pass
        m = row.get("mistake_category")
        if m:
            by_mistake[m] = by_mistake.get(m, 0) + 1

    action_stats = []
    for action, s in by_action.items():
        total = s["hit"] + s["miss"]
        avg = sum(s["returns"]) / len(s["returns"]) if s["returns"] else None
        action_stats.append({
            "action": action,
            "hit": s["hit"],
            "miss": s["miss"],
            "pending": s["pending"],
            "hit_rate": (s["hit"] / total * 100) if total else None,
            "avg_return": avg,
        })
    action_stats.sort(key=lambda x: (x["hit"] + x["miss"]), reverse=True)

    mistake_stats = sorted(
        [{"category": k, "count": v} for k, v in by_mistake.items()],
        key=lambda x: x["count"],
        reverse=True,
    )

    return {"action_stats": action_stats, "mistake_stats": mistake_stats, "total_rows": len(rows)}


def format_memory_for_prompt(
    prior_ticker: List[Dict[str, Any]],
    global_misses: List[Dict[str, Any]],
    mistake_stats: Optional[Dict[str, Any]] = None,
) -> str:
    """Server-side twin of formatMemoryForPrompt in src/journal.js. Output
    is bytewise-compatible so prompts produce equivalent LLM output whether
    assembled client-side or server-side."""
    lines: List[str] = []

    if prior_ticker:
        lines.append("PRIOR ANALYSES FOR THIS TICKER (most recent first):")
        for a in prior_ticker:
            when = str(a.get("created_at") or "")[:10]
            outcome = ""
            if a.get("outcome"):
                rp = a.get("return_pct")
                outcome = f" — {a['outcome']}" + (f" ({rp:.1f}%)" if rp is not None else "")
            price = ""
            if a.get("price_at_analysis") is not None:
                price = f" at {a.get('currency') or ''} {a['price_at_analysis']}"
            lines.append(
                f"- {when}: {a.get('recommendation') or '?'} ({a.get('horizon') or '?'} horizon{price}){outcome}"
            )
            if a.get("mistake_category"):
                msg = f"  mistake: {a['mistake_category']}"
                if a.get("what_was_missed"):
                    msg += f" — {a['what_was_missed']}"
                lines.append(msg)
            if a.get("what_to_check"):
                lines.append(f"  check-next-time: {a['what_to_check']}")
            if a.get("market_regime"):
                lines.append(f"  regime-then: {a['market_regime']}")
            if a.get("lessons"):
                lines.append(f"  free-form lessons: {a['lessons']}")
            reasoning = a.get("reasoning") or {}
            assumptions = reasoning.get("assumptions") if isinstance(reasoning, dict) else None
            if assumptions:
                lines.append(f"  prior assumed: {'; '.join(list(assumptions)[:3])}")

    if global_misses:
        lines.append("")
        lines.append("RECENT WORST CALLS ACROSS TICKERS (avoid these patterns):")
        for m in global_misses:
            when = str(m.get("created_at") or "")[:10]
            cat = f" [{m['mistake_category']}]" if m.get("mistake_category") else ""
            rp = m.get("return_pct")
            note = ""
            if m.get("what_was_missed"):
                note = f" — {m['what_was_missed']}"
            elif m.get("lessons"):
                note = f" — {m['lessons']}"
            lines.append(
                f"- {when} {m.get('ticker')}: {m.get('recommendation')} → "
                f"{rp:.1f}%{cat}{note}" if rp is not None else
                f"- {when} {m.get('ticker')}: {m.get('recommendation')}{cat}{note}"
            )

    if mistake_stats and mistake_stats.get("mistake_stats"):
        lines.append("")
        lines.append("MISTAKE-CATEGORY FREQUENCY (higher = more repeated by this desk):")
        for m in mistake_stats["mistake_stats"][:5]:
            lines.append(f"- {m['category']}: {m['count']} occurrence(s)")

    if not lines:
        return ""

    body = "\n".join(lines)
    return (
        f"\n\n=== MEMORY FROM PAST ANALYSES ===\n{body}\n=== END MEMORY ===\n\n"
        "Before recommending, check whether this thesis matches any pattern above. "
        "In your reasoning, populate priorMistakesConsidered with an array of {pattern, howAvoided} entries. "
        "If the current call closely resembles a prior MISS with no material change in setup, "
        "downgrade the action toward WATCH.\n"
    )

File: src/journal.js
// ============================================================
// ANALYSIS JOURNAL — Supabase-backed structured learning loop
// Every recommendation (stock, portfolio, screen) is saved. Prior calls
// for the same ticker + structured global mistake patterns are injected
// into the next prompt so the model learns from previous errors.
//
// Setup: paste the Project URL + anon key into Settings, then run
// JOURNAL_SETUP_SQL once in Supabase's SQL editor. See PROJECT.md.
// Migration: existing users only need to run JOURNAL_MIGRATION_SQL —
// it adds nullable columns; old rows continue to work unchanged.
// ============================================================
import { createClient } from '@supabase/supabase-js';

// Enumerated mistake categories — keep in sync with UI dropdown + prompt guidance.
export const MISTAKE_CATEGORIES = [
  'stale_data',        // used out-of-date price / news / results
  'wrong_assumption',  // key assumption turned out false
  'valuation_error',   // over/under-paid — multiples were wrong for the regime
  'earnings_surprise', // Q result diverged sharply from expectation
  'macro_shift',       // rate / currency / policy shift invalidated thesis
  'thesis_drift',      // company pivoted, thesis no longer relevant
  'data_gap',          // key datapoint was "unavailable" but call was made anyway
  'concentration',     // portfolio-level: too heavy in one name/sector
  'other',
];

// Bump when the deep-dive prompt schema changes materially. Stored per row
// so we can later filter memory to prompt-versions matching current logic.
export const PROMPT_VERSION = 'v2-2026-07';

export const JOURNAL_SETUP_SQL = `create table if not exists analyses (
  id uuid primary key default gen_random_uuid(),
  created_at timestamptz default now(),
  ticker text not null,
  company_name text,
  provider text,
  horizon text,
  price_at_analysis numeric,
  currency text,
  recommendation text,
  confidence int,
  risk text,
  fair_value text,
  buy_range text,
  target_price numeric,
  summary text,
  reasoning jsonb,
  bull_case jsonb,
  bear_case jsonb,
  full_data jsonb,
  reviewed_at timestamptz,
  price_at_review numeric,
  outcome text,
  return_pct numeric,
  lessons text,
  -- v2 additions (structured learning + audit trail)
  record_type text default 'stock_deepdive',
  mistake_category text,
  what_was_missed text,
  what_to_check text,
  market_regime text,
  model_version text,
  prompt_version text,
  grounded boolean,
  stated_confidence int,
  calibrated_confidence int,
  invalidators jsonb,
  recheck_triggers jsonb,
  prior_mistakes_considered jsonb,
  safety_gate jsonb
);
create index if not exists analyses_ticker_idx on analyses (ticker);
create index if not exists analyses_created_idx on analyses (created_at desc);
create index if not exists analyses_record_type_idx on analyses (record_type);
create index if not exists analyses_mistake_cat_idx on analyses (mistake_category);
alter table analyses disable row level security;`;

// For users upgrading from v1 — additive only.
export const JOURNAL_MIGRATION_SQL = `alter table analyses add column if not exists record_type text default 'stock_deepdive';
alter table analyses add column if not exists mistake_category text;
alter table analyses add column if not exists what_was_missed text;
alter table analyses add column if not exists what_to_check text;
alter table analyses add column if not exists market_regime text;
alter table analyses add column if not exists model_version text;
alter table analyses add column if not exists prompt_version text;
alter table analyses add column if not exists grounded boolean;
alter table analyses add column if not exists stated_confidence int;
alter table analyses add column if not exists calibrated_confidence int;
alter table analyses add column if not exists invalidators jsonb;
alter table analyses add column if not exists recheck_triggers jsonb;
alter table analyses add column if not exists prior_mistakes_considered jsonb;
alter table analyses add column if not exists safety_gate jsonb;
create index if not exists analyses_record_type_idx on analyses (record_type);
create index if not exists analyses_mistake_cat_idx on analyses (mistake_category);`;

function getConfig() {
  try {
    return {
      url: localStorage.getItem('supabaseUrl') || '',
      key: localStorage.getItem('supabaseAnonKey') || '',
    };
  } catch { return { url: '', key: '' }; }
}

let _client = null;
let _clientKey = '';

function client() {
  const { url, key } = getConfig();
  if (!url || !key) return null;
  const cacheKey = `${url}|${key}`;
  if (_client && _clientKey === cacheKey) return _client;
  _client = createClient(url, key, { auth: { persistSession: false } });
  _clientKey = cacheKey;
  return _client;
}

export function journalReady() { return !!client(); }

export async function testConnection() {
  const c = client();
  if (!c) return { ok: false, reason: 'Not configured — add URL and anon key in Settings.' };
  const { count, error } = await c.from('analyses').select('*', { count: 'exact', head: true });
  if (error) return { ok: false, reason: error.message };
  return { ok: true, count: count ?? 0 };
}

function parsePrice(str) {
  if (str == null) return null;
  if (typeof str === 'number') return Number.isFinite(str) ? str : null;
  const cleaned = String(str).replace(/[,\s₹$€£]/g, '').replace(/^Rs\.?/i, '');
  const n = parseFloat(cleaned);
  return Number.isFinite(n) ? n : null;
}

export async function saveAnalysis(record) {
  const c = client();
  if (!c) return { ok: false, reason: 'not-configured' };
  const row = {
    ticker: record.ticker,
    company_name: record.companyName || null,
    provider: record.provider || null,
    horizon: record.horizon || null,
    price_at_analysis: parsePrice(record.priceAtAnalysis),
    currency: record.currency || 'INR',
    recommendation: record.recommendation || null,
    confidence: record.confidence != null ? parseInt(record.confidence, 10) || null : null,
    risk: record.risk || null,
    fair_value: record.fairValue || null,
    buy_range: record.buyRange || null,
    target_price: parsePrice(record.fairValue),
    summary: record.summary || null,
    reasoning: record.reasoning || null,
    bull_case: record.bullCase || null,
    bear_case: record.bearCase || null,
    full_data: record.fullData || null,
    // v2 fields
    record_type: record.recordType || 'stock_deepdive',
    market_regime: record.marketRegime || null,
    model_version: record.modelVersion || null,
    prompt_version: record.promptVersion || PROMPT_VERSION,
    grounded: record.grounded ?? null,
    stated_confidence: record.statedConfidence != null ? parseInt(record.statedConfidence, 10) || null : null,
    calibrated_confidence: record.calibratedConfidence != null ? parseInt(record.calibratedConfidence, 10) || null : null,
    invalidators: record.invalidators || null,
    recheck_triggers: record.recheckTriggers || null,
    prior_mistakes_considered: record.priorMistakesConsidered || null,
    safety_gate: record.safetyGate || null,
  };
  const { data, error } = await c.from('analyses').insert(row).select('id').single();
  if (error) return { ok: false, reason: error.message };
  return { ok: true, id: data.id };
}

export async function listForTicker(ticker, limit = 5) {
  const c = client();
  if (!c || !ticker) return [];
  const { data, error } = await c
    .from('analyses')
    .select('*')
    .eq('ticker', ticker.toUpperCase())
    .order('created_at', { ascending: false })
    .limit(limit);
  if (error) { console.warn('journal listForTicker:', error.message); return []; }
  return data || [];
}

export async function listRecentMisses(limit = 5) {
  const c = client();
  if (!c) return [];
  const { data, error } = await c
    .from('analyses')
    .select('*')
    .eq('outcome', 'MISS')
    .order('return_pct', { ascending: true })
    .limit(limit);
  if (error) { console.warn('journal listRecentMisses:', error.message); return []; }
  return data || [];
}

export async function listAll(limit = 100, recordType) {
  const c = client();
  if (!c) return [];
  let q = c.from('analyses').select('*').order('created_at', { ascending: false }).limit(limit);
  if (recordType) q = q.eq('record_type', recordType);
  const { data, error } = await q;
  if (error) { console.warn('journal listAll:', error.message); return []; }
  return data || [];
}

export async function updateReview(id, patch) {
  const c = client();
  if (!c) return { ok: false, reason: 'not-configured' };
  const row = {
    reviewed_at: new Date().toISOString(),
    price_at_review: patch.priceAtReview != null ? parsePrice(patch.priceAtReview) : undefined,
    outcome: patch.outcome || undefined,
    return_pct: patch.returnPct != null ? Number(patch.returnPct) : undefined,
    lessons: patch.lessons != null ? patch.lessons : undefined,
    mistake_category: patch.mistakeCategory !== undefined ? (patch.mistakeCategory || null) : undefined,
    what_was_missed: patch.whatWasMissed !== undefined ? (patch.whatWasMissed || null) : undefined,
    what_to_check: patch.whatToCheck !== undefined ? (patch.whatToCheck || null) : undefined,
    market_regime: patch.marketRegime !== undefined ? (patch.marketRegime || null) : undefined,
  };
  Object.keys(row).forEach(k => row[k] === undefined && delete row[k]);
  const { error } = await c.from('analyses').update(row).eq('id', id);
  if (error) return { ok: false, reason: error.message };
  return { ok: true };
}

export async function deleteAnalysis(id) {
  const c = client();
  if (!c) return { ok: false, reason: 'not-configured' };
  const { error } = await c.from('analyses').delete().eq('id', id);
  if (error) return { ok: false, reason: error.message };
  return { ok: true };
}

// Aggregate stats: hit rate per action, worst recurring mistake categories, count by outcome.
// Runs client-side over recent rows — fine for a personal single-user app.
export async function getMistakeStats(limit = 500) {
  const c = client();
  if (!c) return null;
  const { data, error } = await c
    .from('analyses')
    .select('recommendation,outcome,return_pct,mistake_category,record_type')
    .order('created_at', { ascending: false })
    .limit(limit);
  if (error) return null;
  const rows = data || [];
  const byAction = {};   // { BUY: {hit, miss, pending, avgReturn} }
  const byMistake = {};  // { mistake_category: count }
  for (const r of rows) {
    const a = r.recommendation || 'UNKNOWN';
    if (!byAction[a]) byAction[a] = { hit: 0, miss: 0, pending: 0, returns: [] };
    if (r.outcome === 'HIT')  byAction[a].hit++;
    else if (r.outcome === 'MISS') byAction[a].miss++;
    else byAction[a].pending++;
    if (r.return_pct != null) byAction[a].returns.push(Number(r.return_pct));
    if (r.mistake_category) byMistake[r.mistake_category] = (byMistake[r.mistake_category] || 0) + 1;
  }
  const actionStats = Object.entries(byAction).map(([action, s]) => {
    const total = s.hit + s.miss;
    const avgReturn = s.returns.length ? s.returns.reduce((a, b) => a + b, 0) / s.returns.length : null;
    return {
      action,
      hit: s.hit,
      miss: s.miss,
      pending: s.pending,
      hitRate: total > 0 ? (s.hit / total) * 100 : null,
      avgReturn,
    };
  }).sort((a, b) => (b.hit + b.miss) - (a.hit + a.miss));
  const mistakeStats = Object.entries(byMistake)
    .map(([category, count]) => ({ category, count }))
    .sort((a, b) => b.count - a.count);
  return { actionStats, mistakeStats, totalRows: rows.length };
}

// Return recent MISS rows for a specific ticker — used by the safety gate
// to detect repeat-mistake risk (same thesis, same failure pattern).
export async function listMissesForTicker(ticker, limit = 5) {
  const c = client();
  if (!c || !ticker) return [];
  const { data, error } = await c
    .from('analyses')
    .select('*')
    .eq('ticker', ticker.toUpperCase())
    .eq('outcome', 'MISS')
    .order('created_at', { ascending: false })
    .limit(limit);
  if (error) return [];
  return data || [];
}

// Build a rich context block. Now includes:
//  - Prior calls for this ticker (with outcome + structured mistake fields)
//  - Global mistake-pattern frequencies so the model knows what to watch for
//  - Explicit instruction to reference and avoid repeat patterns
export function formatMemoryForPrompt(priorTicker, globalMisses, mistakeStats) {
  const lines = [];
  if (priorTicker?.length) {
    lines.push('PRIOR ANALYSES FOR THIS TICKER (most recent first):');
    priorTicker.forEach(a => {
      const when = new Date(a.created_at).toISOString().slice(0, 10);
      const outcome = a.outcome ? ` — ${a.outcome}${a.return_pct != null ? ` (${a.return_pct.toFixed(1)}%)` : ''}` : '';
      const price = a.price_at_analysis != null ? ` at ${a.currency || ''} ${a.price_at_analysis}` : '';
      lines.push(`- ${when}: ${a.recommendation || '?'} (${a.horizon || '?'} horizon${price})${outcome}`);
      if (a.mistake_category)  lines.push(`  mistake: ${a.mistake_category}${a.what_was_missed ? ` — ${a.what_was_missed}` : ''}`);
      if (a.what_to_check)     lines.push(`  check-next-time: ${a.what_to_check}`);
      if (a.market_regime)     lines.push(`  regime-then: ${a.market_regime}`);
      if (a.lessons)           lines.push(`  free-form lessons: ${a.lessons}`);
      if (a.reasoning?.assumptions?.length) {
        lines.push(`  prior assumed: ${a.reasoning.assumptions.slice(0, 3).join('; ')}`);
      }
    });
  }
  if (globalMisses?.length) {
    lines.push('');
    lines.push('RECENT WORST CALLS ACROSS TICKERS (avoid these patterns):');
    globalMisses.forEach(m => {
      const when = new Date(m.created_at).toISOString().slice(0, 10);
      const cat = m.mistake_category ? ` [${m.mistake_category}]` : '';
      lines.push(`- ${when} ${m.ticker}: ${m.recommendation} → ${m.return_pct?.toFixed(1)}%${cat}${m.what_was_missed ? ` — ${m.what_was_missed}` : (m.lessons ? ` — ${m.lessons}` : '')}`);
    });
  }
  if (mistakeStats?.mistakeStats?.length) {
    lines.push('');
    lines.push('MISTAKE-CATEGORY FREQUENCY (higher = more repeated by this desk):');
    mistakeStats.mistakeStats.slice(0, 5).forEach(m => {
      lines.push(`- ${m.category}: ${m.count} occurrence(s)`);
    });
  }
  if (!lines.length) return '';
  return `\n\n=== MEMORY FROM PAST ANALYSES ===\n${lines.join('\n')}\n=== END MEMORY ===\n\nBefore recommending, check whether this thesis matches any pattern above. In your reasoning, populate priorMistakesConsidered with an array of {pattern, howAvoided} entries. If the current call closely resembles a prior MISS with no material change in setup, downgrade the action toward WATCH.\n`;
}
